game/plugin/skills/woodcutting/src/woodcutting.plugin.kts

```diff
@@ -0,0 +1,117 @@
+
+        return obj
+    }
+
+    fun isSuccessful(mob: Player): Boolean {
+        return rand(100) <= wood.chance * 100
+    }
+}
+
+class WoodcuttingAction(val player: Player, val tool: Axe, val target: WoodcuttingTarget) : AsyncDistancedAction<Player>(PULSES, true, player, target.position, TREE_SIZE) {
+
+    companion object {
+        private val PULSES = 0
+        private val TREE_SIZE = 1;
+
+        fun axeFor(player: Player): Axe? {
+            return AXES
+                    .filter { it.level <= player.skills.woodcutting.currentLevel }
+                    .filter { player.equipment.contains(it.id) || player.inventory.contains(it.id) }
+                    .sortedByDescending { it.level }
+                    .firstOrNull()
+        }
+
+        /**
+         * Starts a [WoodcuttingAction] for the specified [Player], terminating the [Message] that triggered it.
+         */
+        fun start(message: ObjectActionMessage, player: Player, wood: Wood) {
+            val axe = axeFor(player)
+            if (axe != null) {
+                val action = WoodcuttingAction(player, axe, WoodcuttingTarget(message.id, message.position, wood))
+                System.out.println("Starting action")
+                player.startAction(action)
+            } else {
+                player.sendMessage("You do not have a axe for which you have the level to use.")
+            }
+
+            message.terminate()
+        }
+    }
+
+    override fun start() {
+        System.out.println("Action start")
+        mob.turnTo(position)
+
+        val level = mob.skills.woodcutting.currentLevel
+
+        if (level < target.wood.level) {
+            mob.sendMessage("You do not have the required level to cut down this tree.")
+            stop()
+        }
+    }
+
+    override suspend fun executeActionAsync() {
+        System.out.println("Action execute")
+        mob.sendMessage("You swing your axe at the tree.")
+        mob.playAnimation(tool.animation)
+
+        wait(tool.pulses)
+        //
+
+        val obj = target.getObject(mob.world)
+        if (!obj.isPresent) {
+            System.out.println("Cannot find object!")
+            stop()
+            return
+        }
+
+        if (mob.inventory.freeSlots() == 0) {
+            mob.inventory.forceCapacityExceeded()
+            stop()
+            return
+        }
+        if (mob.inventory.add(target.wood.id)) {
+            //TODO: Use lookup from utils once it has a lookup function for IDs
+            val woodName = ItemDefinition.lookup(target.wood.id).name.toLowerCase();
+            mob.sendMessage("You managed to cut some $woodName.")
+            mob.skillSet.addExperience(Skill.WOODCUTTING, target.wood.exp)
+        }
+
+        if (target.isSuccessful(mob)) {
+            //respawn time: http://runescape.wikia.com/wiki/Trees
+            val respawn = ((30 * 1000) / 600) + ((rand(150) * 1000) / 600) // between 30 sec and 3 min respawm
```

Comment:
```kotlin
val respawn = TimeUnit.SECONDS.toMillis(30 + rand(150)) / GameConstants.PULSE_DELAY
```

using `java.util.concurrent.TimeUnit` and `org.apollo.game.GameConstants`.
Might also be nice to put the `30` into its own constant - say, `MINIMUM_RESPAWN_TIME`.